File: learning-api/src/courses/getCourseListByDomain.mjs
import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
import { DynamoDBDocumentClient, QueryCommand } from "@aws-sdk/lib-dynamodb";

const client = new DynamoDBClient();
const docClient = DynamoDBDocumentClient.from(client);

export const handler = async (event) => {
  // get course domain from path parameters
  const courseDomain = event.pathParameters.courseDomain;

  // set command parameters for query
  let params = {
    TableName: process.env.COURSES_TABLE,
    IndexName: 'courseDomainIndex',
    KeyConditionExpression: 'courseDomain = :courseDomain',
    ExpressionAttributeValues: { ':courseDomain': courseDomain },
    ScanIndexForward: false // sort by createdAt (descending)
  };

  try {
    // query data from DynamoDB
    const command = new QueryCommand(params);
    const response = await docClient.send(command);

    // sort the list by createdAt (descending)
    let items = response.Items;
    items.sort((a, b) => new Date(b.createdAt) - new Date(a.createdAt));

    return {
      statusCode: 200,
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(items)
    };
  }
  catch (err) {
    console.error(err);
    return {
      statusCode: 500,
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ message: 'Failed to get course list by domain' })
    };
  }
};

File: learning-api-python/src/courses/get_course_list_by_domain.py
import boto3
import os
import json

def handler(event, context):
    # get DynamoDB table
    dynamodb = boto3.resource("dynamodb")
    table = dynamodb.Table(os.getenv("COURSES_TABLE"))

    # get course domain from path parameters
    courseDomain = event.get("pathParameters", {}).get("courseDomain")

    try:
        # query data from DynamoDB
        response = table.query(
            IndexName="courseDomainIndex",
            KeyConditionExpression="courseDomain = :courseDomain",
            ExpressionAttributeValues={":courseDomain": courseDomain},
            ScanIndexForward=False # sort by createdAt in descending order
        )
        # get course list from response
        courses = response.get("Items", [])
        # convert Decimal to float
        for course in courses:
            course["rating"] = float(course["rating"])

        response = {
            "statusCode": 200,
            "headers": { "Content-Type": "application/json" },
            "body": json.dumps(courses)
        }
    except Exception as e:
        print(f'Error: {e}') # log the error
        response = {
            "statusCode": 500,
            "headers": { "Content-Type": "application/json" },
            "body": json.dumps({"message": "Failed to get course list"})
        }
    return response